---
title: "Exposome Data Integration with Omic Data"
author: "Carles Hernandez-Ferer and Juan R. Gonzalez"
date: "`r doc_date()`"
package: "`r pkg_ver('omicRexposome')`"
abstract: >
  This is an introductory guide to integration analysis between exposome and omics data with R package omicRexposome. The document ilustrates two types of analysis: 1) Association analysis, that are performed between exposome and a single omic data-set; and 2) Integration analysis where multiple data-sets, including exposome data, are analysed at the same time.
vignette: >
  %\VignetteIndexEntry{Exposome Data Integration with Omic Data}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
output: 
  BiocStyle::html_document
---

```{r setup, include=FALSE}
BiocStyle::markdown()
knitr::opts_chunk$set(echo = TRUE, warnings=FALSE)
```

```{r load_omicRexposome, message=FALSE}
library(omicRexposome)
```

```{r load_exposome, message=FALSE}
library(rexposome)

path <- paste0(path.package("rexposome"), .Platform$file.sep, "extdata")
description <- paste0(path, .Platform$file.sep, "description.csv")
phenotype <- paste0(path, .Platform$file.sep, "phenotypes.csv")
exposures <- paste0(path, .Platform$file.sep, "exposures.csv")

exp <- read_exposome(
    exposures = exposures, 
    description = description, 
    phenotype = phenotype,
    exposures.samCol = 1, 
    description.expCol = 2, 
    description.famCol = 1, 
    phenotype.samCol = 1
)
```

# Association Studies

## Methylation - Transcriptome Data Associaction

```{r methy_data, cache=FALSE}
data("methy_r", package="omicRexposome")
```

```{r methy_analysis, message=FALSE, warning=FALSE}
methy <- assocES(exp, methy_r, ~sex+age, 
    select=c("Pb", "Co", "THM", "G_pesticides"))
```

```{r methy_tables}
cbind(
    Hits=tableHits(methy, th=0.001),
    Lambda=tableLambda(methy)[ , -1]
)
```

```{r methy_plot_tables}
gridExtra::grid.arrange(
    plotHits(methy, th=0.001) + ggplot2::ggtitle("Methilation Hits"),
    plotLambda(methy) + ggplot2::ggtitle("Methylation Lambdas"),
    nrow=2
)
```

```{r methy_plot_volcano}
gridExtra::grid.arrange(
    plotAssociation(methy, rid="Pb", type="volcano", tPV=-log10(5e-05)) + 
        ggplot2::ggtitle("Transcriptome - Pb Association"),
    plotAssociation(methy, rid="THM", type="volcano", tPV=-log10(5e-05)) + 
        ggplot2::ggtitle("Transcriptome - THM Association"),
    ncol=2
)
```

## Exposome - Transcriptome Data Associaction

The aim of this analysis is to perform an association test between the gene expression levels and the exposures. To this end, a `MultiDataSet` with an `ExposomeSet` and an `ExpressionSet` is required.

```{r gexp_data, cache=FALSE}
data("gexp_r", package="omicRexposome")
```

The association analysis between exposures and gene expression is done using the method `assocGE`. This method has the same behavior than `assocME`.

```{r gexp_analysis, message=FALSE, warning=FALSE}
gexp <- assocES(exp, gexp_r, ~sex+age, 
    select=c("Pb", "Co", "THM", "G_pesticides"))
```

```{r gexp_tables}
cbind(
    Hits=tableHits(gexp, th=0.001),
    Lambda=tableLambda(gexp)[ , -1]
)
```

```{r gexp_plot_tables}
gridExtra::grid.arrange(
    plotHits(gexp, th=0.001) + ggplot2::ggtitle("Transcriptome Hist"),
    plotLambda(gexp) + ggplot2::ggtitle("Transcriptome Lambdas"),
    nrow=2
)
```

```{r gexp_plot_volcano}
gridExtra::grid.arrange(
    plotAssociation(gexp, rid="Pb", type="volcano", tFC=2, tPV=-log10(5e-05)) + 
        ggplot2::ggtitle("Transcriptome - Pb Association"),
    plotAssociation(gexp, rid="THM", type="volcano", tFC=2, tPV=-log10(5e-05)) + 
        ggplot2::ggtitle("Transcriptome - THM Association"),
    ncol=2
)
```

## Exposome - Transcriptome Data Associaction

The aim of this analysis is to perform an association test between proteome levels and the exposures. To this end, a `MultiDataSet` with an `ExposomeSet` and an `ExpressionSet` is required.

```{r prot_data, cache=FALSE}
data("prot_r", package="omicRexposome")
```

The association analysis between exposures and gene expression is done using the method `assocGE`. This method has the same behavior than `assocME`.

```{r prot_analysis, message=FALSE, warning=FALSE}
gexp <- assocES(exp, prot_r, ~sex+age, 
    select=c("Pb", "Co", "THM", "G_pesticides"))
```

```{r prot_tables}
cbind(
    Hits=tableHits(gexp, th=0.001),
    Lambda=tableLambda(gexp)[ , -1]
)
```

```{r prot_plot_tables}
gridExtra::grid.arrange(
    plotHits(gexp, th=0.001) + ggplot2::ggtitle("Proteome Hist"),
    plotLambda(gexp) + ggplot2::ggtitle("Proteome Lambdas"),
    nrow=2
)
```

```{r prot_plot_volcano}
gridExtra::grid.arrange(
    plotAssociation(gexp, rid="Pb", type="volcano", tFC=2, tPV=-log10(5e-05)) + 
        ggplot2::ggtitle("Proteome - Pb Association"),
    plotAssociation(gexp, rid="THM", type="volcano", tFC=2, tPV=-log10(5e-05)) + 
        ggplot2::ggtitle("Proteome - THM Association"),
    ncol=2
)
```

# Integration Analysis

```{r impute_Exp, echo=FALSE}
exp <- impute(exp)
```

```{r crossomics_mcca, message=FALSE, warning=FALSE, results='hide'}
cr_mcca <- crossomics(
    list(exposome=exp, transcriptome=gexp_r, methylome=methy_r), 
    method = "mcca"
)
cr_mcca
```

```{r crossomics_mcia, message=FALSE, warning=FALSE}
cr_mcia <- crossomics(
    list(exposome=exp, transcriptome=gexp_r, methylome=methy_r), 
    method = "mcia"
)
cr_mcia
```

```{r plot_mcca, messages=FALSE, warnings=FALSE}
plotIntegration(cr_mcca)
```

```{r plot_mcia, messages=FALSE, warnings=FALSE}
plotIntegration(cr_mcia)
```
