---
title: "Exposome Data Integration with Omic Data"
author: "Carles Hernandez-Ferer and Juan R. Gonzalez"
date: "`r doc_date()`"
package: "`r pkg_ver('rexposome')`"
abstract: >
  This is an introductory guide to integration analysis between exposome and omics data with R package rexposome. The document ilustrates two types of analysis: 1) Association analysis, that are performed between exposome and a single omic data-set; and 2) Integration analysis where multiple data-sets, including exposome data, are analysed at the same time.
vignette: >
  %\VignetteIndexEntry{Bioconductor style for HTML documents}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
output: 
  BiocStyle::html_document
---

```{r setup, include=FALSE}
BiocStyle::markdown()
knitr::opts_chunk$set(echo = TRUE, warnings=FALSE)
```

# Introduction

## Pipeline

## Data Format

## Installation

```{r github, eval=FALSE}
devtools::install_github("isglobal-brge/omicRexposome")
```

The methods provided by `rexposome` and used in this document relies in other existing Bioconductor packages. All association and integration methods requires a `r Biocpkg("MultiDataSet")` object. The following are the R packages used in association analyses:

 * `r Biocpkg("snpStats")`
 * `r Biocpkg("limma")`

These are the packages used in the integration analyses:

 * `r Biocpkg("omicade4")`
 * `r CRANpkg("pma")`

# Analysis

The data used for in this vignete has two differens sources:

 * `rexposome` - exposome data will be used
 * `BRGEdata` - gene expresion and methylation data will be used

Under this sections two types of analysis are explained: __Association Analysis__ and __Integration Analysis__.

Both types of analysis requests a `MultiDataSet` object. While __Association Analysis__ are performed using a `MultiDataSet` object that contains an `ExposomeSet` plus a `SnpSet`, `MethylationSet` or `ExpressionSet`, __Integration Analysis__ needs a `MultiDataSet` object with, at last, two of these objects.

We start loading `omicRexposome`:

```{r load_omicRexposome, message=FALSE}
library(omicRexposome)
```

The *exposome* data is loaded from `rexposome` package as it is illustrated in its vignettes:

```{r load_exposome, message=FALSE}
library(rexposome)

path <- paste0(path.package("rexposome"), .Platform$file.sep, "extdata")
description <- paste0(path, .Platform$file.sep, "description.csv")
phenotype <- paste0(path, .Platform$file.sep, "phenotypes.csv")
exposures <- paste0(path, .Platform$file.sep, "exposures.csv")

exp <- read_exposome(
    exposures = exposures, 
    description = description, 
    phenotype = phenotype,
    exposures.samCol = 1, 
    description.expCol = 2, 
    description.famCol = 1, 
    phenotype.samCol = 1
)
```

## Associaction Analysis

Then we load the `MultiDataSet` package to be able to create the input objects for association methods.

```{r load_mds, message=FALSE}
library(MultiDataSet)
```

### Exposome - Methylome Data Associaction

The goal of this analysis is to perform an EWAS between the methylation levels and the exposures. To this end, a `MultiDataSet` with an `ExposomeSet` and a `MethylationSet` is required.

```{r methy_data, cache=FALSE}
data("mset_rd", package = "BRGEdata")
```

Once the data is loaded a `MultiDataSet` object is created containing both data-sets:

```{r methy_mds, message=FALSE, warning=FALSE}
mds <- createMultiDataSet()
mds <- add_methy(mds, mset_r)
mds <- add_exp(mds, exp)
```

After creating the `MultiDataSet` created the common samples between sets are kept and the others discarded:

```{r methy_mds_common, error=TRUE}
mds <- commonSamples(mds)
```

```{r methy_mds_show}
mds
```

After obtaining the common samples, the association analysis between CpGs and exposures is done using `assocME`. This method requires a `MultiDataSet` object and a formula with the covariates. The optional argument `select` allows to perform the association on a sub-set of exposures (if `set = "exposures`) or sub-set of _outcomes_(if `set = "phenotypes"`).

```{r methy_analysis, message=FALSE, warning=FALSE}
ewas <- assocME(mds, ~sex+age, 
    select=c("Pb", "Co", "THM", "G_pesticides"))
```

The result is an object of class `ResultSet`:

```{r methy_show}
ewas
```

The raw results of the analysis can be obtained using the method `extract` and giving it the exposures of interest. If no exposure (or phenotype) is given, `extract` will return a table with the results for all the exposures (or phenotypes).

```{r methy_rid}
rid(ewas)
```

```{r methy_extract}
head(topTable(ewas, rid="G_pesticides")[ , 1:8], n=3)
```

The `ResultSet` can be used to plot the already seen QQ and Manhattan plots. Also a volcano plot.

```{r methy_volcano, fig.width=6, fig.height=6}
plotAssociation(ewas, rid="G_pesticides", type="volcano")
```

```{r methy_rm, echo=FALSE}
rm("mds", "ewas")
```

```{r message=FALSE, echo=FALSE, include=FALSE}
gc()
```

The same analysis can be done using an `ExposomeClust` instead of an `ExposomeSet`. 

```{r methy_cls, cache=FALSE}
data("eclust", package = "rexposome")
```

```{r show=FALSE}
sampleNames(expo_c) <- gsub("SIMSAB", "id", sampleNames(expo_c))
```

```{r methy_mds2, message=FALSE, warning=FALSE}
mds <- createMultiDataSet()
mds <- add_methy(mds, mset_r)
mds <- add_cls(mds, expo_c)
```

After creating the `MultiDataSet` the method `assocME` can be called directly. It will look for the common samples before running the association test.

```{r methy_analysis2, message=FALSE, warning=FALSE}
methy <- assocME(mds, ~sex+age)
methy
```

The results of the association analysis can be obtained using the tag `"cluster"` in the argument `rid` of `extract` method:

```{r methy_extract2}
head(topTable(methy, rid="cluster")[ , 1:8], n=3)
```

And a Manhattan plot showing the association between the clustering and the protein levels of expression can be drawn:

```{r methy_manhattan2}
plotAssociation(methy, rid="cluster", type="volcano")
```

```{r methy_rm2, echo=FALSE}
rm("mset_r", "mds", "methy")
```

### Exposome - Transcriptome Data Associaction

The aim of this analysis is to perform an association test between the gene expression levels and the exposures. To this end, a `MultiDataSet` with an `ExposomeSet` and an `ExpressionSet` is required.

```{r gexp_data, cache=FALSE}
data("gexp_rd", package="BRGEdata")
```


Once the data is loaded a `MultiDataSet` object is created containing both data-sets:

```{r gexp_mds, message=FALSE, warning=FALSE}
mds <- createMultiDataSet()
mds <- add_genexp(mds, gexp_r)
mds <- add_exp(mds, exp)
```

After creating the `MultiDataSet` created the common samples between sets are kept and the others discarded:

```{r gexp_mds_common, error=TRUE}
mds <- commonSamples(mds)
```

```{r gexp_mds_show}
mds
```

The association analysis between exposures and gene expression is done using the method `assocGE`. This method has the same behavior than `assocME`.

```{r gexp_analysis, message=FALSE, warning=FALSE}
gexp <- assocGE(mds, ~sex+age, 
    select=c("Pb", "Co", "THM", "G_pesticides"))
```

The obtained object is also a `ResultSet`. The raw table can be obtained using `extract`:

```{r gexp_extract}
tt <- topTable(gexp)
table(tt$exposure)
head(tt[ , c("logFC", "P.Value", "adj.P.Val", "exposure")], n=5)
```

And all three QQ, Manhattan and volcano plots can be dawn using `plotAssociation`:

```{r gexp_volcano, fig.width=6, fig.height=6}
plotAssociation(gexp, rid="Pb", type="volcano",
    tPV=-log(10e-3), tFC=2)
```

```{r gexp_rm, echo=FALSE}
rm("mds", "gexp", "tt")
```

The same analysis can be done using an `ExposomeClust` instead of an `ExposomeSet`. 

```{r gexp_mds2, message=FALSE, warning=FALSE}
mds <- createMultiDataSet()
mds <- add_genexp(mds, gexp_r)
mds <- add_cls(mds, expo_c)
```

After creating the `MultiDataSet` the method `assocGE` can be called directly. It will look for the common samples before running the association test.

```{r gexp_analysis2, message=FALSE, warning=FALSE}
gexp <- assocGE(mds, ~sex+age)
gexp
```

The results of the association analysis can be obtained using the tag `"cluster"` in the argument `rid` of `extract` method:

```{r gexp_extract2}
head(topTable(gexp, rid="cluster"), n=3)
```

And a Manhattan plot showing the association between the clustering and the protein levels of expression can be drawn:

```{r gexp_manhattan2}
plotAssociation(gexp, rid="cluster", type="manhattan")
```

```{r gexp_rm2, echo=FALSE}
rm("gexp_r", "mds", "gexp")
```


## Integration Analysis

The aim on the integration analysis is to analyze multiple data-sets in a single shot, looking for similar behaviors in terms of features and through the data-sets. `rexposome` allows to perform this analysis using two strategies: _multiple co-inertia analysis_ (`mcia`) and _sparse multiple canonical correlation analysis_ (`mcca`).

To perform an integration analysis using exposome, transcriptome and proteome data a `MultiDataSet` object is required with the thee data-sets.

```{r}
data("exposome", package = "rexposome")
ldBRGE("gexp")
## ldBRGE("prot")
```

```{r gexp_samples2, message=FALSE, warning=FALSE}
sampleNames(gexp4) <- gsub("id", "SIMSAB", sampleNames(gexp4))
```

All the data-sets needs to be complete. So the exposures should be imputed:

```{r impute_exposures, cache=TRUE}
expo <- impute(expo)
```

```{r crs_mds, message=FALSE, warning=FALSE}
mds <- createMultiDataSet()
mds <- add_prot(mds, prot4)
mds <- add_genexp(mds, gexp4)
mds <- add_exp(mds, expo)
```

Once the data-sets is created, we look for the common samples.

```{r crs_mds_common, error=TRUE}
mds <- commonSamples(mds)
mds
```

Both integration analysis are done using the method `crossomics`. The argument `method` can takes the values `"mcia"` and `"mcca"`.

```{r crs_analysis1, warning=FALSE, message=FALSE}
crs1 <- crossomics(mds, method="mcia")
```

```{r crs_analysis2, warning=FALSE, message=FALSE}
crs2 <- crossomics(mds, method="mcca")
```

The method `crossomics` returns a `ResultSet` that can be used in the method `plotIntegration` to plot the results.


```{r crs_plot1, fig.width=12, fig.height=6}
plotIntegration(crs1)
```

```{r crs_plot2, fig.width=12, fig.height=10}
plotIntegration(crs2, lb.th=0.3)
```
